design/baremetal-operator/sub-states.md

```diff
@@ -0,0 +1,155 @@
+a BMH lifecycle it is.
+
+## Motivation
+
+The primary motivation is to enable operators an easier insight into what
+is going on with their nodes. This simplifies debugging and makes people feel
+better about long-running processes. It may also be helpful for UI tools
+to be able to display some sort of progress information.
+
+### Goals
+
+* Provide a way to know what action is being executed right now without
+  accessing Ironic API.
+
+### Non-Goals
+
+* Provide an access to the Ironic Node `provision_state` field.
+* Establish a direct one-to-one mapping between provision states and the
+  new field values.
+* Provide more granularity than what the Ironic API can report.
+* The values of the new field will not be a part of the API contract and can
+  change as a result of backend changes.
+* Define a non-empty value for each possible situation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator reasonably familiar with the provisioning process, I would like
+to know what action is being executed on a BMH right now.
+
+## Design Details
+
+* The `ProvisionStatus` structure is extended with a new string field
+  `CurrentActivity`.
+* The value of the field is re-calculated based on the result described below
+  on each reconciliation.
+
+### Implementation Details/Notes/Constraints
+
+The new field will be populated based on the following rules, where `Node` is
+an Ironic Node object. The rules are defined in pseudo-code.
+
+This proposal introduces a concept of "cleaning" with a narrower definition
+that in Ironic. While in Ironic "cleaning" refers to any pre-deployment
+actions, in our case "cleaning" will be used in the context of "cleaning
+disks".
+
+* IF `Node.provision_state` in "cleaning", "clean wait":
+
+  * IF `Node.clean_step` is null
+    AND `Node.target_provision_state` is "available":
+
+    Activity is `Preparing to clean disks`.
+
+  * IF `Node.clean_step` is null
+    AND `Node.target_provision_state` is "manageable":
+
+    Activity is `Preparing to configure host`.
+
+  * IF `Node.clean_step["interface"]` == "raid":
+
+    Activity is `Configuring RAID`.
+
+  * IF `Node.clean_step["interface"]` == "bios":
+
+    Activity is `Configuring firmware settings`.
+
+  * IF `Node.clean_step["step"]` in "erase\_devices",
+    "erase\_devices\_metadata":
+
+    Activity is `Cleaning disks`.
+
+* IF `Node.provision_state` in "deploying", "deploy call-back":
+
+  * IF `Node.deploy_step` is null OR `Node.deploy_step["step"]` is "deploy":
+
+    Activity is `Preparing to write image`.
+
+  * IF `Node.deploy_step["step"]` is "write\_image":
+
+    Activity is `Writing image`.
+
+  * IF `Node.deploy_step["step"]` is "prepare\_instance\_boot":
+
+    Activity is `Configuring boot`.
+
+  * IF `Node.deploy_step["step"]` in "tear\_down\_agent",
+    "switch\_to\_tenant\_network", "boot\_instance":
+
+    Activity is `Rebooting into operating system`.
+
```

Comment:
Probably a default case could be useful to empty `CurrentActivity` when none of the above conditions match, otherwise the field could report an incorrect situation?

Reply:
It's implicit in my proposal, I'll make it explicit.